src/piece_picker.cpp

```diff
@@ -2923,18 +2923,24 @@ namespace libtorrent
 
 	bool piece_picker::is_piece_ready(int index) const
 	{
-		if (!is_piece_finished(index))
-			return false;
+		TORRENT_ASSERT(index < int(m_piece_map.size()));
+		TORRENT_ASSERT(index >= 0);
 
 		piece_pos const& p = m_piece_map[index];
 		int state = p.download_queue();
 
-		if (state < piece_pos::num_download_categories)
+		if (p.index != piece_pos::we_have_index && state == piece_pos::piece_open)
+		{
+			for (int i = 0; i < piece_pos::num_download_categories; ++i)
+				TORRENT_ASSERT(find_dl_piece(i, index) == m_downloads[i].end());
```

Comment:
I think it always feels scary to have an assert as the only statement in a loop that doesn't have braces. Imagine if this assert statement would simply disappear in release builds, then the return statement would end up in the loop.

I think you could probably just remove the loop and the assert, to keep in simple.